UserReplicaSetManager - Libs + Mad Dog

hareeshnagaraj opened the pull request:
### Description
_What is the purpose of this PR? What is the current behavior? New behavior? Relevant links (e.g. Trello) and/or information pertaining to PR?_

Libs changes and mad-dog tests associated with this PR - https://github.com/AudiusProject/audius-protocol/pull/1058

Back-compat release of libs for both pre- and post-contract deploy support



### Tests
_List the manual tests and repro instructions to verify that this PR works as anticipated. Include log analysis if possible.\
:exclamation: If this change impacts clients, make sure that you have tested the clients :exclamation:_

New mad dog test added and passing in CI


**:exclamation: Reminder :bulb::exclamation::**\
If this PR touches a critical flow (such as Indexing, Uploads, Gateway or the Filesystem), make sure to add the `requires-special-attention` label. **Add relevant labels as necessary.**

hareeshnagaraj commented:
![](https://github.trello.services/images/mini-trello-icon.png) [UserReplicaSetManager Contract Indexing + Libs](https://trello.com/c/LVhFZuj7/1723-userreplicasetmanager-contract-indexing-libs)

hareeshnagaraj commented on discovery-provider/Dockerfile:
```diff
@@ -1,5 +1,5 @@
 # requires authentication to Dockerhub
-FROM audius/contracts:1.0.0 as contracts
+FROM audius/contracts:user-replica-set-manager-12-16-20 as contracts
```
just publish 2.0.0, v safe to do so

hareeshnagaraj replied:
consult w/others before doing this

hareeshnagaraj replied:
per discussion w @dmanjunath setting to v2.0.0-rc1

hareeshnagaraj replied:
https://hub.docker.com/layers/audius/contracts/v2.0.0-rc1/images/sha256-e5d74d8fafcce0cdc7ed165dfaa2a9a9a5a5ab8ffd3144f8c72cd2f8da5033f5?context=repo

hareeshnagaraj commented on discovery-provider/src/tasks/index.py:
```diff
@@ -133,6 +134,24 @@ def fetch_tx_receipts(self, block_transactions):
         raise Exception(f"index.py | fetch_tx_receipts Expected ${num_submitted_txs} received {num_processed_txs}")
     return block_tx_with_receipts
 
+def update_user_replica_set_manager_address_if_necessary(self):
```
add comment explaining why this is necessary

hareeshnagaraj commented on discovery-provider/src/tasks/index.py:
```diff
@@ -183,21 +204,22 @@ def index_blocks(self, db, blocks_list):
             # Parse tx events in each block
             for tx in sorted_txs:
                 tx_hash = web3.toHex(tx["hash"])
+                logger.error(tx)
```
rm

hareeshnagaraj commented on libs/src/services/creatorNode/CreatorNodeSelection.js:
```diff
@@ -63,7 +77,15 @@ class CreatorNodeSelection extends ServiceSelection {
     console.info('CreatorNodeSelection - final decision tree state', this.decisionTree)
-    return { primary, secondaries, services: servicesMap }
+    console.log(primary)
+    console.log(contentNodeEndpointToSpID[primary])
```
remove excess logs

hareeshnagaraj commented on discovery-provider/src/__init__.py:
```diff
@@ -97,14 +99,23 @@ def initContracts():
+    user_replica_set_manager_inst = web3.eth.contract(
+        address=user_replica_set_manager_address, abi=abi_values["UserReplicaSetManager"]["abi"]
+    )
+    logger.error(f"Found address for USRM: {user_replica_set_manager_address}")
```
remove this

hareeshnagaraj commented on discovery-provider/src/queries/get_l2_cnode.py:
```diff
@@ -0,0 +1,20 @@
+from src.utils.db_session import get_db_read_replica
+from src.queries.query_helpers import paginate_query
+
+def get_l2_cnodes():
```
are filters necessary here?

hareeshnagaraj replied:
i kind of think no? as in querying for a subset etc. pagination should likely cover the requirements

hareeshnagaraj commented on discovery-provider/src/tasks/user_replica_set.py:
```diff
@@ -0,0 +1,229 @@
+        primary
+    ).call()
+    logger.error(primary_cn_endpoint_info)
+    logger.error(f"PRIMARY CN INFO {primary_cn_endpoint_info}")
```
rm these logs

hareeshnagaraj commented on discovery-provider/src/tasks/user_replica_set.py:
```diff
@@ -0,0 +1,229 @@
+        shared_config["eth_contracts"]["registry"]
+    )
+    logger.error(f"USER REPLICA SET PY: ETH REGISTRY {eth_registry_address}")
+    eth_registry_instance = eth_web3.eth.contract(
```
TODO - Enable a cache from index_network_peers - this would mean both storing values in `index_network_peers.py` retrieving them here and falling back if necessary

hareeshnagaraj commented on libs/src/api/user.js:
```diff
@@ -502,12 +520,15 @@ class Users extends Base {
     let metadata = { ...newMetadata }
     exclude.map(excludedKey => delete metadata[excludedKey])
 
+    console.log(metadata)
```
remove this

hareeshnagaraj commented on libs/src/services/creatorNode/index.js:
```diff
@@ -319,6 +319,10 @@ class CreatorNode {
       const status = await axios(req)
+      console.log(`latest from req`)
+      console.log(status.data.latestBlockNumber)
+      console.log(`latest from user: `)
```
remove excess logs

hareeshnagaraj commented on libs/src/services/dataContracts/userReplicaSetManagerClient.js:
```diff
@@ -0,0 +1,67 @@
+    return tx
+  }
+
+  // TODO: AddOrUpdateContentNode Functionality
```
IMO the best time to add this is when we do cross-cnode auth but I can add it now also

hareeshnagaraj replied:
this is done! local sandbox cmd validates functionality ready for content node integration

hareeshnagaraj commented on libs/src/services/dataContracts/userReplicaSetManagerClient.js:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  async getContentNodeWallet (spId) {
+    const method = await this.getMethod('getContentNodeWallet', spId)
```
use getWalletAddressString here as well

hareeshnagaraj commented on libs/src/api/user.js:
```diff
@@ -316,14 +318,21 @@ class Users extends Base {
     const oldMetadata = { ...user }
 
+    console.log(`Updating from ${oldMetadata.creator_node_endpoint}`)
+    console.log(`Updating to ${newMetadata.creator_node_endpoint}`)
```
remove these logs and cleanup changes in this file

hareeshnagaraj commented on libs/src/services/dataContracts/userReplicaSetManagerClient.js:
```diff
@@ -0,0 +1,123 @@
+const signatureSchemas = require('../../../data-contracts/signatureSchemas')
+const Web3Manager = require('../web3Manager/index')
+
+class UserReplicaSetManagerClient extends ContractClient {
```
Comments throughout prior to marking as 'ready for review'

hareeshnagaraj commented on discovery-provider/src/tasks/index.py:
```diff
@@ -147,6 +170,7 @@ def index_blocks(self, db, blocks_list):
         logger.info(
             f"index.py | index_blocks | {self.request.id} | block {block.number} - {block_index}/{num_blocks}"
         )
+        logger.error(f"index.py - KNOWN ADDRS {contract_addresses}")
```
remove this

hareeshnagaraj commented on discovery-provider/src/tasks/user_replica_set.py:
```diff
@@ -0,0 +1,259 @@
+# In the case of failure where a given user has valid primary/secondaries but no 
+# endpoint string, a client must fetch the endpoint associated with ID 
+# Indexing CANNOT block on this endpoint string optimization.
+def get_endpoint_string_from_sp_ids(
```
cc @SidSethi @dmanjunath I want to discuss a potential retry pattern here but non-blocking functionally

SidSethi replied:
yeah lets talk about this

hareeshnagaraj commented:
Unit test cleanup remaining but mainline code is ready for review

raymondjacobson commented on discovery-provider/scripts/lint.py:
```diff
@@ -12,5 +12,5 @@ def lint_directory(directory):
-lint_directory("src/")
-lint_directory("tests/")
+lint_directory("src/queries/")
+# lint_directory("tests/")
```
just putting a comment so we don't forget to put this back

raymondjacobson commented on discovery-provider/src/models.py:
```diff
@@ -153,6 +152,8 @@ class User(Base):
     updated_at = Column(DateTime, nullable=False)
     created_at = Column(DateTime, nullable=False)
+    primary = Column(Integer, nullable=True)
+    secondaries = Column(postgresql.ARRAY(Integer), nullable=True)
```
just a thought -- do you think we will ever want to store more metadata w/ secondaries? if so maybe a json structure is better?

hareeshnagaraj replied:
i don't think so? as in it should always be an array of IDs, even if we switch away from POA

raymondjacobson commented on discovery-provider/src/queries/queries.py:
```diff
@@ -550,3 +551,11 @@ def get_creator_node_users():
         return api_helpers.error_response(str(e), 400)
+
+@bp.route("/l2/content_node", methods=("GET",))
+def get_l2_content_nodes():
```
i would say here and throughout, we should probably not hard-code "l2" into the nomenclature here. this could exist on any chain, right? why do we want to call it l2?

hareeshnagaraj replied:
zero pref here, open to anything

raymondjacobson replied:
🤷  get_content_nodes?

SidSethi replied:
yeah my main concern was to avoid POA but also, given that primarily disco indexes L2 i can see just omitting that term and only using L1 in the minority case eg index_network_peers
nbd tho

hareeshnagaraj replied:
how would we feel about making it usrm_cnode or something to indicate it's as part of UserReplicaSetManager? bc L2 is also wrong if we migrate to a diff. chain like it might be 2 distinct L1 chains not eth +L2

hareeshnagaraj replied:
because that's really what this is - the current view of content nodes on UserReplicaSEtManager wherever that may live

dmanjunath replied:
i think being specific about POA is not bad fwiw. like just having a poa specific endpoint and functions (as well as whatever chain we move to) will probably help us when thinking about backwards compat in the future

raymondjacobson commented on libs/initScripts/helpers/spRegistration.js:
```diff
@@ -70,14 +70,19 @@ async function deregisterLocalService (audiusLibs, serviceType, serviceEndpoint)
  * @param {Object} audiusLibs fully formed audius libs instance with eth contracts connection
  */
-async function queryLocalServices (audiusLibs, serviceTypeList) {
+async function queryLocalServices (audiusLibs, serviceTypeList, usrmLibs = null) {
```
maybe add some @param comments and explain why we need another libs?

raymondjacobson commented on libs/src/api/user.js:
```diff
@@ -567,6 +573,33 @@ class Users extends Base {
   }
+
+  async _updateReplicaSet (userId, metadata) {
+    let primaryEndpoint = CreatorNode.getPrimary(metadata['creator_node_endpoint'])
```
nit, metadata.creator_node_endpoint in js land

raymondjacobson reviewed (commented):
libs stuff looks pretty good to me, i think the get SP id call is probably fine. Will want to be cautious rolling out & test very thoroughly on staging w/ live ammo. Will do another pass after sid & dheeraj take a look since they were closer to the design :)

cheran-senthil commented on discovery-provider/src/tasks/user_replica_set.py:
```diff
@@ -0,0 +1,267 @@
+                secondary_endpoint = secondary_info[1]
+
+            if secondary_endpoint:
+                endpoint_string = "{},{}".format(endpoint_string, secondary_endpoint)
```
We are using both `.format` and `fstrings` in our code. We might want to move towards `fstrings` as the newer standard for formatting in Python.

hareeshnagaraj replied:
out of scope

cheran-senthil commented on libs/initScripts/local.js:
```diff
@@ -149,6 +196,149 @@ const run = async () => {
+const addL2ContentNode = async (
+  audiusLibs,
+  ethAccounts,
+  newCnodeId,
+  newCnodeDelegateWallet
```
I noticed that we renamed this variable at some other parts, might be nice to rename it everywhere moving forward.

cheran-senthil commented on discovery-provider/src/__init__.py:
```diff
@@ -139,6 +150,7 @@ def create_celery(test_config=None):
     global playlist_factory
     global user_library_factory
     global ipld_blacklist_factory
+    global user_replica_set_manager
     global contract_addresses
```
There are a lot of global variables here and on line 137 above, I think it might be nice to merge these into a single line.

hareeshnagaraj replied:
i think this is unrelated / out of scope but i agree

cheran-senthil commented on discovery-provider/src/models.py:
```diff
@@ -603,3 +606,24 @@ def __repr__(self):
+
+    PrimaryKeyConstraint(is_current, cnode_id, blockhash)
+
+    def __repr__(self):
+        return f"<L2ContentNode(blockhash={self.blockhash},\
+is_current={self.is_current},cnode_id={self.cnode_id},\
+delegate_owner_wallet={self.delegate_owner_wallet}, proposer_sp_ids={self.proposer_sp_ids},\
+proposer_1_address={self.proposer_1_address},proposer_2_address={self.proposer_2_address},\
+proposer_3_address={self.proposer_3_address})>"
```
I think we can use string addition or using `"""` to escape the entire string. This happens in other places too.

cheran-senthil reviewed (commented):
Not pressing, but I think it might be nice to enforce some kind of linting standard and/or use a formatter like `yapf` or `black` for the python code in the repository.

Most of my comments below are Python linting related nits, and should not be blocking for a merge.

vicky-g commented on discovery-provider/alembic/versions/5aaa6198def8_user_primary_secondary_columns.py:
```diff
@@ -0,0 +1,30 @@
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.add_column('users', sa.Column('primary', sa.Integer(), nullable=True))
```
is it intentional to leave these col values as nullable?

vicky-g replied:
/ is there going to be a retroactive fill for primaries and secondaries for all users?

hareeshnagaraj replied:
yes these must be nullable for backwards compatibility during the interim where users have yet to upgrade

vicky-g commented on discovery-provider/alembic/versions/5dd6a55bb738_poa_content_nodes.py:
```diff
@@ -0,0 +1,39 @@
+        sa.Column('blockhash', sa.String(), nullable=False),
+        sa.Column('created_at', sa.DateTime(), nullable=False),
+        sa.Column('is_current', sa.Boolean(), nullable=False),
+        sa.Column('cnode_id', sa.Integer(), nullable=False),
```
jw, what is the cnode_id? / maybe add some comments to what these fields are?

hareeshnagaraj replied:
https://content-node.audius.co/health_check if you look here, every node of any type within the protocol has a dedicated spID

vicky-g commented on discovery-provider/src/queries/queries.py:
```diff
@@ -550,3 +551,11 @@ def get_creator_node_users():
+        cnodes = get_l2_cnodes()
+        return api_helpers.success_response(cnodes)
+    except exceptions.ArgumentError as e:
+        return api_helpers.error_response(str(e), 400)
```
maybe we should make this except block generalized to take any type of error?

hareeshnagaraj replied:
i don't feel strongly

vicky-g replied:
well I more so thinking that if an exception of not type `ArgumentError` was thrown, then the except block wouldn't be able to catch that error. then the dn wouldnt be able to serve an api error response

vicky-g commented on discovery-provider/src/tasks/index.py:
```diff
@@ -133,6 +134,28 @@ def fetch_tx_receipts(self, block_transactions):
     return block_tx_with_receipts
 
+# During each indexing iteration, check if the address for UserReplicaSetManager
+# Has been set in the L2 contract registry - if so, update the global contract_addresses object
```
nit - lowercase has

vicky-g commented on discovery-provider/src/tasks/user_replica_set.py:
```diff
@@ -0,0 +1,267 @@
+                cnode_id = args._cnodeId if "_cnodeId" in args else None
+
+                # if the user id is not in the lookup object, it hasn't been initialized yet
+                # first, get the user object from the db(if exists or create a new one)
```
nit - space ` db (if exists or create`

vicky-g commented on discovery-provider/src/tasks/user_replica_set.py:
```diff
@@ -0,0 +1,267 @@
+                    ret_user = lookup_user_record(update_task, session, entry, block_number, block_timestamp)
+                    user_replica_set_events_lookup[user_id] = {"user": ret_user, "events": []}
+
+                # def lookup_poa_cnode_record(self, update_task, session, entry, block_number, block_timestamp):
```
?

vicky-g commented on discovery-provider/src/tasks/user_replica_set.py:
```diff
@@ -0,0 +1,267 @@
+    cnode_id = event_args._cnodeId
+
+    cnode_record_exists = session.query(L2ContentNode).filter_by(cnode_id=cnode_id).count() > 0
+    logger.error(f"lookup_poa_cnode_record | {cnode_id} record exists={cnode_record_exists}")
```
logger.info?

vicky-g commented on discovery-provider/src/tasks/user_replica_set.py:
```diff
@@ -0,0 +1,267 @@
+    return sp_factory_inst
+
+def parse_poa_cnode_record(self, update_task, session, entry, cnode_record):
+    event_args = entry["args"]
```
nit - I feel like we can direcftly assign cnode_record to have these fields, instead of creating these vars, and then updating cnode_record

hareeshnagaraj replied:
i think this is better for clarity tbh, good comment but I'm going to stick with the current format

vicky-g commented on discovery-provider/src/tasks/users.py:
```diff
@@ -230,3 +236,13 @@ def get_metadata_overrides_from_ipfs(session, update_task, user_record):
+    secondary_replica_set_configured = user_record.secondaries is not None 
+    if secondary_replica_set_configured:
+        secondary_replica_set_configured = len(user_record.secondaries) > 0
+    upgraded_replica_set = primary_replica_set_configured or secondary_replica_set_configured
```
dont we want `primary_replica_set_configured and secondary_replica_set_configured`? what if one upgraded but the other didnt?

vicky-g replied:
should we error? or handle this somehow

hareeshnagaraj replied:
i don't think so - how would that scenario occur? maybe take a look at how we interact with UserReplicaSetManager and the function signatures for some context

vicky-g commented on libs/initScripts/local.js:
```diff
@@ -137,6 +142,48 @@ const run = async () => {
         break
+
+      case 'update-user-replica-set':
+        console.log(`Usage: node local.js update-user-replica-set userId=1 primary=2 secondaries=3,1`)
```
nit - maybe make this a comment instead of a log

hareeshnagaraj replied:
i actually think its better as a log - that way it spits out the usage every time

vicky-g commented on libs/src/services/creatorNode/CreatorNodeSelection.js:
```diff
@@ -71,7 +84,6 @@ class CreatorNodeSelection extends ServiceSelection {
       val: { primary, secondaries: secondaries.toString(), services: Object.keys(servicesMap).toString() }
     })
 
-    console.info('CreatorNodeSelection - final decision tree state', this.decisionTree)
```
is there a reason for this removal?

hareeshnagaraj replied:
No good catch - will add this back in

SidSethi commented on discovery-provider/src/tasks/user_replica_set.py:
```diff
@@ -0,0 +1,267 @@
+    )
+    return sp_factory_inst
+
+def parse_poa_cnode_record(self, update_task, session, entry, cnode_record):
```
none of these params except `cnode_record` are used, can remove

hareeshnagaraj replied:
entry is used repeatedly...? you can see it right below this line

SidSethi commented on discovery-provider/src/tasks/user_replica_set.py:
```diff
@@ -0,0 +1,267 @@
+        endpoint_string = "{}".format(primary_endpoint)
+        for secondary_id in secondaries:
+            secondary_endpoint = None
+            secondary_cache_key = get_sp_id_key(secondary_id)
+            secondary_info_cached = get_pickled_key(update_task.redis, secondary_cache_key)
+            if secondary_info_cached:
+                secondary_endpoint = secondary_info_cached[1]
+                logger.info(f"user_replica_set.py | CACHE HIT FOR {secondary_cache_key}, found {secondary_info_cached}")
+
+            if not secondary_endpoint:
+                logger.info(
+                    f"user_replica_set.py | CACHE MISS FOR {secondary_cache_key}, found {secondary_info_cached}"
+                )
+                if sp_factory_inst is None:
+                    sp_factory_inst = get_sp_factory_inst(self, update_task)
+                secondary_info = sp_factory_inst.functions.getServiceEndpointInfo(
+                    content_node_service_type,
+                    secondary_id
+                ).call()
+                secondary_endpoint = secondary_info[1]
```
this block of code is repeated from lines 154-162. could instead make an array `spIDs = [primary, ...secondaries]` and then call this once in a for loop over that

hareeshnagaraj replied:
i think that actually makes the code worse - I considered your suggestion but decided to explicitly separate the logic for primary vs secondaries. i do agree and think there is a better abstraction though to avoid code repetition, will update

hareeshnagaraj replied:
fixed

SidSethi reviewed (commented):
comments from the indexing side - overall looks great, as we've discussed

hareeshnagaraj commented on identity-service/scripts/dev-server.sh:
```diff
@@ -2,7 +2,7 @@
 set -e
 
-link_libs=false
+link_libs=true
```
i think it should be merged

hareeshnagaraj commented on libs/src/api/user.js:
```diff
@@ -686,6 +691,47 @@ class Users extends Base {
+
+  async _updateReplicaSet (userId, metadata) {
+    // Attempt to update through UserReplicaSetManagerClient if present  
+    if (!this.contracts.UserReplicaSetManagerClient) {
```
@dmanjunath @SidSethi what happens now is an attempt to initialize this new client and if that fails - fall back to the old UserFactoryClient

hareeshnagaraj commented on libs/src/services/contracts/ContractClient.js:
```diff
@@ -59,6 +59,10 @@ class ContractClient {
     try {
       if (!this._contractAddress) {
         this._contractAddress = await this.getRegistryAddress(this.contractRegistryKey)
+        if (this._contractAddress === "0x0000000000000000000000000000000000000000") {
```
Without this check, we would consider a contract to be 'initialized' at the zero address

hareeshnagaraj reviewed (commented):
More self review comments

hareeshnagaraj commented on discovery-provider/src/tasks/user_replica_set.py:
```diff
@@ -0,0 +1,267 @@
+                args = entry["args"]
+                # Check if _userId is present
+                # If user id is found in the event args, update the local lookup object
+                user_id = args._userId if "_userId" in args else None
+                user_ids.add(user_id)
```
seems like adding None has no effect right? bc thsi is used to determine whether or not to refresh cached values for a given set of users and the value is only checked if num_user_replica_set_changes >0. personally don't see an issue w/leaving this as it is exactly right

dmanjunath replied:
it has no effect functionally but if i understand correctly, the set of user_ids is returned and plumbed all the way to set and retrieve cached users from redis. So the redis would have a key with `user:id:None` (from here https://github.com/AudiusProject/audius-protocol/blob/master/discovery-provider/src/utils/redis_cache.py#L1310). 

One day in the future if we were debugging something and we saw a random None key in there I'd probably be thrown off and be like wtf is this a bug? I would opt to do something like
```
if user_id:
  user_ids.add(user_id)
```

hareeshnagaraj commented on discovery-provider/src/tasks/user_replica_set.py:
```diff
@@ -0,0 +1,262 @@
+
+# Initializes sp_factory if necessary and retrieves spID
+# Returns initialized instance of contract and endpoint
+def get_endpoint_from_id(self, update_task, sp_factory_inst, sp_id):
```
if for whatever reason this ID is completely invalid on the eth-contracts, continue and only set null for the invalid id (ie endpoint1,,endpoint4)

hareeshnagaraj commented on discovery-provider/alembic/versions/5aaa6198def8_user_primary_secondary_columns.py:
```diff
@@ -0,0 +1,30 @@
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.add_column('users', sa.Column('primary', sa.Integer(), nullable=True))
+    op.add_column('users', sa.Column('secondaries', sa.ARRAY(sa.Integer()), nullable=True))
```
i agree but i'm torn on the exact format - we can finalize tmrw. primaryID, secondaryIDs? or primarySpID, secondarySpIDs you get the gist

hareeshnagaraj commented on discovery-provider/src/tasks/user_replica_set.py:
```diff
@@ -0,0 +1,262 @@
+    cache_key = get_sp_id_key(sp_id)
+    # Attempt to fetch from cache
+    sp_info_cached = get_pickled_key(update_task.redis, cache_key)
+    if sp_info_cached:
```
from 2/1/21 - move the endpoint reconstruction into the query layer instead of indexing

hareeshnagaraj commented:
play through the scenario where a non-existent spID is written - unlikely but let's confirm the behavior

hareeshnagaraj commented:
primary_id, secondary_ids _everywhere_

hareeshnagaraj commented:
Mad-dog test output

```
ubuntu@ip-172-31-29-163:~/audius-protocol/mad-dog$ npm run start test-usrm 10

> mad-dog@0.0.1 start /home/ubuntu/audius-protocol/mad-dog
> node src/index.js --inspect "test-usrm" "10"

Config dir: /home/ubuntu/.audius
/home/ubuntu/.audius/config.json
/home/ubuntu/.audius/eth-config.json
2021-02-04T23:54:14.707Z info: 	🐶 * Woof Woof * Welcome to Mad-Dog 🐶
2021-02-04T23:54:14.714Z info: 	Running test [userReplicaSetManager]
Selected discprov http://audius-disc-prov_web-server_1:5000 [ { stage: 'Check Short Circuit', val: null },
  { stage: 'Get All Services',
    val: [ 'http://audius-disc-prov_web-server_1:5000' ] },
  { stage: 'Filter Out Known Unhealthy',
    val: [ 'http://audius-disc-prov_web-server_1:5000' ] },
  { stage: 'Get Selection Round',
    val: [ 'http://audius-disc-prov_web-server_1:5000' ] },
  { stage: 'Raced And Found Best',
    val: 'http://audius-disc-prov_web-server_1:5000' },
  { stage: 'Made A Selection',
    val: 'http://audius-disc-prov_web-server_1:5000' } ]
Selected discprov http://audius-disc-prov_web-server_1:5000 [ { stage: 'Check Short Circuit', val: null },
  { stage: 'Get All Services',
    val: [ 'http://audius-disc-prov_web-server_1:5000' ] },
  { stage: 'Filter Out Known Unhealthy',
    val: [ 'http://audius-disc-prov_web-server_1:5000' ] },
  { stage: 'Get Selection Round',
    val: [ 'http://audius-disc-prov_web-server_1:5000' ] },
  { stage: 'Raced And Found Best',
    val: 'http://audius-disc-prov_web-server_1:5000' },
  { stage: 'Made A Selection',
    val: 'http://audius-disc-prov_web-server_1:5000' } ]
Selected discprov http://audius-disc-prov_web-server_1:5000 [ { stage: 'Check Short Circuit', val: null },
  { stage: 'Get All Services',
    val: [ 'http://audius-disc-prov_web-server_1:5000' ] },
  { stage: 'Filter Out Known Unhealthy',
    val: [ 'http://audius-disc-prov_web-server_1:5000' ] },
  { stage: 'Get Selection Round',
    val: [ 'http://audius-disc-prov_web-server_1:5000' ] },
  { stage: 'Raced And Found Best',
    val: 'http://audius-disc-prov_web-server_1:5000' },
  { stage: 'Made A Selection',
    val: 'http://audius-disc-prov_web-server_1:5000' } ]
Selected discprov http://audius-disc-prov_web-server_1:5000 [ { stage: 'Check Short Circuit', val: null },
  { stage: 'Get All Services',
    val: [ 'http://audius-disc-prov_web-server_1:5000' ] },
  { stage: 'Filter Out Known Unhealthy',
    val: [ 'http://audius-disc-prov_web-server_1:5000' ] },
  { stage: 'Get Selection Round',
    val: [ 'http://audius-disc-prov_web-server_1:5000' ] },
  { stage: 'Raced And Found Best',
    val: 'http://audius-disc-prov_web-server_1:5000' },
  { stage: 'Made A Selection',
    val: 'http://audius-disc-prov_web-server_1:5000' } ]
2021-02-04T23:54:15.888Z info: 	Starting: [Add users]...
2021-02-04T23:54:15.943Z info: 	Creating new user...
2021-02-04T23:54:16.025Z info: 	Creating new user...
2021-02-04T23:54:16.060Z info: 	Creating new user...
2021-02-04T23:54:16.092Z info: 	Creating new user...
CreatorNodeSelection - final decision tree state [ { stage: 'Get All Services',
    val:
     [ 'http://cn1_creator-node_1:4000',
       'http://cn2_creator-node_1:4001',
       'http://cn3_creator-node_1:4002' ] },
  { stage: 'Filter To Whitelist',
    val:
     [ 'http://cn1_creator-node_1:4000',
       'http://cn2_creator-node_1:4001',
       'http://cn3_creator-node_1:4002' ] },
  { stage: 'Filter From Blacklist',
    val:
     [ 'http://cn1_creator-node_1:4000',
       'http://cn2_creator-node_1:4001',
       'http://cn3_creator-node_1:4002' ] },
  { stage: 'Filter Out Unhealthy And Outdated',
    val:
     [ 'http://cn1_creator-node_1:4000',
       'http://cn2_creator-node_1:4001',
       'http://cn3_creator-node_1:4002' ] },
  { stage: 'Select Primary And Secondaries',
    val:
     { primary: 'http://cn1_creator-node_1:4000',
       secondaries:
        'http://cn2_creator-node_1:4001,http://cn3_creator-node_1:4002',
       services:
        'http://cn1_creator-node_1:4000,http://cn2_creator-node_1:4001,http://cn3_creator-node_1:4002' } } ]
CreatorNodeSelection - final decision tree state [ { stage: 'Get All Services',
    val:
     [ 'http://cn1_creator-node_1:4000',
       'http://cn2_creator-node_1:4001',
       'http://cn3_creator-node_1:4002' ] },
  { stage: 'Filter To Whitelist',
    val:
     [ 'http://cn1_creator-node_1:4000',
       'http://cn2_creator-node_1:4001',
       'http://cn3_creator-node_1:4002' ] },
  { stage: 'Filter From Blacklist',
    val:
     [ 'http://cn1_creator-node_1:4000',
       'http://cn2_creator-node_1:4001',
       'http://cn3_creator-node_1:4002' ] },
  { stage: 'Filter Out Unhealthy And Outdated',
    val:
     [ 'http://cn1_creator-node_1:4000',
       'http://cn2_creator-node_1:4001',
       'http://cn3_creator-node_1:4002' ] },
  { stage: 'Select Primary And Secondaries',
    val:
     { primary: 'http://cn3_creator-node_1:4002',
       secondaries:
        'http://cn1_creator-node_1:4000,http://cn2_creator-node_1:4001',
       services:
        'http://cn3_creator-node_1:4002,http://cn1_creator-node_1:4000,http://cn2_creator-node_1:4001' } } ]
CreatorNodeSelection - final decision tree state [ { stage: 'Get All Services',
    val:
     [ 'http://cn1_creator-node_1:4000',
       'http://cn2_creator-node_1:4001',
       'http://cn3_creator-node_1:4002' ] },
  { stage: 'Filter To Whitelist',
    val:
     [ 'http://cn1_creator-node_1:4000',
       'http://cn2_creator-node_1:4001',
       'http://cn3_creator-node_1:4002' ] },
  { stage: 'Filter From Blacklist',
    val:
     [ 'http://cn1_creator-node_1:4000',
       'http://cn2_creator-node_1:4001',
       'http://cn3_creator-node_1:4002' ] },
  { stage: 'Filter Out Unhealthy And Outdated',
    val:
     [ 'http://cn1_creator-node_1:4000',
       'http://cn2_creator-node_1:4001',
       'http://cn3_creator-node_1:4002' ] },
  { stage: 'Select Primary And Secondaries',
    val:
     { primary: 'http://cn2_creator-node_1:4001',
       secondaries:
        'http://cn3_creator-node_1:4002,http://cn1_creator-node_1:4000',
       services:
        'http://cn2_creator-node_1:4001,http://cn3_creator-node_1:4002,http://cn1_creator-node_1:4000' } } ]
CreatorNodeSelection - final decision tree state [ { stage: 'Get All Services',
    val:
     [ 'http://cn1_creator-node_1:4000',
       'http://cn2_creator-node_1:4001',
       'http://cn3_creator-node_1:4002' ] },
  { stage: 'Filter To Whitelist',
    val:
     [ 'http://cn1_creator-node_1:4000',
       'http://cn2_creator-node_1:4001',
       'http://cn3_creator-node_1:4002' ] },
  { stage: 'Filter From Blacklist',
    val:
     [ 'http://cn1_creator-node_1:4000',
       'http://cn2_creator-node_1:4001',
       'http://cn3_creator-node_1:4002' ] },
  { stage: 'Filter Out Unhealthy And Outdated',
    val:
     [ 'http://cn1_creator-node_1:4000',
       'http://cn2_creator-node_1:4001',
       'http://cn3_creator-node_1:4002' ] },
  { stage: 'Select Primary And Secondaries',
    val:
     { primary: 'http://cn2_creator-node_1:4001',
       secondaries:
        'http://cn3_creator-node_1:4002,http://cn1_creator-node_1:4000',
       services:
        'http://cn2_creator-node_1:4001,http://cn3_creator-node_1:4002,http://cn1_creator-node_1:4000' } } ]
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 19]. Retrying request at attempt #0...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 19]. Retrying request at attempt #0...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 19]. Retrying request at attempt #1...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 19]. Retrying request at attempt #0...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 19]. Retrying request at attempt #1...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 19]. Retrying request at attempt #0...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 19]. Retrying request at attempt #2...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 19]. Retrying request at attempt #1...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 19]. Retrying request at attempt #2...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 19]. Retrying request at attempt #1...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 19]. Retrying request at attempt #3...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 19]. Retrying request at attempt #2...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 19]. Retrying request at attempt #3...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 19]. Retrying request at attempt #2...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 19]. Retrying request at attempt #4...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 18]. Retrying request at attempt #4...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 18]. Retrying request at attempt #3...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 18]. Retrying request at attempt #3...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 16]. Retrying request at attempt #5...
Attempted max retries with endpoint http://audius-disc-prov_web-server_1:5000
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 16]. Retrying request at attempt #5...
Attempted max retries with endpoint http://audius-disc-prov_web-server_1:5000
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 16]. Retrying request at attempt #4...
http://audius-disc-prov_web-server_1:5000 is too far behind [block diff: 16]. Retrying request at attempt #4...
Selected discprov null [ { stage: 'Check Short Circuit', val: null },
  { stage: 'Get All Services',
    val: [ 'http://audius-disc-prov_web-server_1:5000' ] },
  { stage: 'Filter Out Known Unhealthy', val: [] },
  { stage: 'Get Selection Round', val: [] },
  { stage: 'No Services Left To Try' },
  { stage: 'Failed Everything -- Resetting' } ]
Error: All Discovery Providers are unhealthy and unavailable.
    at DiscoveryProvider.getHealthyDiscoveryProviderEndpoint (/home/ubuntu/audius-protocol/libs/src/services/discoveryProvider/index.js:557:13)
    at process._tickCallback (internal/process/next_tick.js:68:7)
Selected discprov null [ { stage: 'Check Short Circuit', val: null },
  { stage: 'Get All Services',
    val: [ 'http://audius-disc-prov_web-server_1:5000' ] },
  { stage: 'Filter Out Known Unhealthy', val: [] },
  { stage: 'Get Selection Round', val: [] },
  { stage: 'No Services Left To Try' },
  { stage: 'Failed Everything -- Resetting' } ]
Error: All Discovery Providers are unhealthy and unavailable.
    at DiscoveryProvider.getHealthyDiscoveryProviderEndpoint (/home/ubuntu/audius-protocol/libs/src/services/discoveryProvider/index.js:557:13)
    at process._tickCallback (internal/process/next_tick.js:68:7)
[Block Check] Waiting for #60 to be indexed...
[Block Check] Discovery Node has indexed #60!
Uploading file to http://cn3_creator-node_1:4002/image_upload
[Block Check] Waiting for #60 to be indexed...
[Block Check] Discovery Node has indexed #60!
[Block Check] Waiting for #60 to be indexed...
[Block Check] Waiting for #60 to be indexed...
Uploading file to http://cn2_creator-node_1:4001/image_upload
[Block Check] Discovery Node has indexed #60!
[Block Check] Discovery Node has indexed #60!
Uploading file to http://cn1_creator-node_1:4000/image_upload
Uploading file to http://cn2_creator-node_1:4001/image_upload
2021-02-04T23:54:42.374Z info: 	Created new user=2
[Block Check] Waiting for #72 to be indexed...
[Block Check] Discovery Node has indexed #72!
2021-02-04T23:54:44.646Z info: 	Added users, userIds=2
2021-02-04T23:54:45.542Z info: 	Created new user=1
[Block Check] Waiting for #72 to be indexed...
[Block Check] Discovery Node has indexed #72!
2021-02-04T23:54:45.567Z info: 	Added users, userIds=2,1
2021-02-04T23:54:47.659Z info: 	Created new user=3
[Block Check] Waiting for #72 to be indexed...
[Block Check] Discovery Node has indexed #72!
2021-02-04T23:54:47.717Z info: 	Added users, userIds=2,1,3
2021-02-04T23:54:47.905Z info: 	Created new user=4
[Block Check] Waiting for #72 to be indexed...
[Block Check] Discovery Node has indexed #72!
2021-02-04T23:54:47.929Z info: 	Added users, userIds=2,1,3,4
2021-02-04T23:54:47.929Z info: 	Finished: [Add users].
2021-02-04T23:54:47.930Z info: 	Starting: [Upgrade to creator]...
2021-02-04T23:54:55.774Z info: 	Finished upgrading creator for user=1
[Block Check] Waiting for #80 to be indexed...
[Block Check] Discovery Node has indexed #80!
2021-02-04T23:54:56.185Z info: 	Finished upgrading creator for user=4
[Block Check] Waiting for #80 to be indexed...
[Block Check] Discovery Node has indexed #80!
2021-02-04T23:54:56.335Z info: 	Finished upgrading creator for user=3
2021-02-04T23:54:56.345Z info: 	Finished upgrading creator for user=2
[Block Check] Waiting for #80 to be indexed...
[Block Check] Waiting for #80 to be indexed...
[Block Check] Discovery Node has indexed #80!
[Block Check] Discovery Node has indexed #80!
2021-02-04T23:54:56.404Z info: 	Finished: [Upgrade to creator].
2021-02-04T23:54:56.665Z info: 	Validating content-nodes on UserReplicaSetManager
2021-02-04T23:54:56.760Z info: 	Found UserReplicaSetManager and Discovery Provider match for spID=1, delegateWallet=0xc992b2bB49B447C4a405b0A8efE590cAbB59368D
2021-02-04T23:54:56.789Z info: 	Found UserReplicaSetManager and ServiceProviderFactory match for spID=1, delegateWallet=0xc992b2bB49B447C4a405b0A8efE590cAbB59368D
2021-02-04T23:54:56.827Z info: 	Found UserReplicaSetManager and Discovery Provider match for spID=2, delegateWallet=0x5B68e00F7bEB980548F52e5298E182c3A77f35b1
2021-02-04T23:54:56.858Z info: 	Found UserReplicaSetManager and Discovery Provider match for spID=3, delegateWallet=0x59422AB3CC4a2436f6Cc3B3800790438cF1e005E
2021-02-04T23:54:56.878Z info: 	Found UserReplicaSetManager and ServiceProviderFactory match for spID=2, delegateWallet=0x5B68e00F7bEB980548F52e5298E182c3A77f35b1
2021-02-04T23:54:56.899Z info: 	Found UserReplicaSetManager and ServiceProviderFactory match for spID=3, delegateWallet=0x59422AB3CC4a2436f6Cc3B3800790438cF1e005E
2021-02-04T23:54:56.899Z info: 	Finished validating content-nodes on UserReplicaSetManager
2021-02-04T23:54:57.017Z info: 	userId: 1 Replica Set Info: 1, 2,3
2021-02-04T23:54:57.017Z info: 	userId: 1 Replica Set String: http://cn1_creator-node_1:4000,http://cn2_creator-node_1:4001,http://cn3_creator-node_1:4002
2021-02-04T23:54:57.017Z info: 	userId: 1 primaryID: 1 primaryIdFromEndointStr: 1
2021-02-04T23:54:57.018Z info: 	userId: 1 secondaryId: 2 secondaryIdFromEndpointStr: 2
2021-02-04T23:54:57.018Z info: 	userId: 1 secondaryId: 3 secondaryIdFromEndpointStr: 3
2021-02-04T23:54:57.068Z info: 	userId: 2 Replica Set Info: 3, 1,2
2021-02-04T23:54:57.068Z info: 	userId: 2 Replica Set String: http://cn3_creator-node_1:4002,http://cn1_creator-node_1:4000,http://cn2_creator-node_1:4001
2021-02-04T23:54:57.068Z info: 	userId: 2 primaryID: 3 primaryIdFromEndointStr: 3
2021-02-04T23:54:57.068Z info: 	userId: 2 secondaryId: 1 secondaryIdFromEndpointStr: 1
2021-02-04T23:54:57.068Z info: 	userId: 2 secondaryId: 2 secondaryIdFromEndpointStr: 2
2021-02-04T23:54:57.098Z info: 	userId: 3 Replica Set Info: 2, 3,1
2021-02-04T23:54:57.099Z info: 	userId: 3 Replica Set String: http://cn2_creator-node_1:4001,http://cn3_creator-node_1:4002,http://cn1_creator-node_1:4000
2021-02-04T23:54:57.099Z info: 	userId: 3 primaryID: 2 primaryIdFromEndointStr: 2
2021-02-04T23:54:57.099Z info: 	userId: 3 secondaryId: 3 secondaryIdFromEndpointStr: 3
2021-02-04T23:54:57.099Z info: 	userId: 3 secondaryId: 1 secondaryIdFromEndpointStr: 1
2021-02-04T23:54:57.104Z info: 	userId: 4 Replica Set Info: 2, 3,1
2021-02-04T23:54:57.104Z info: 	userId: 4 Replica Set String: http://cn2_creator-node_1:4001,http://cn3_creator-node_1:4002,http://cn1_creator-node_1:4000
2021-02-04T23:54:57.104Z info: 	userId: 4 primaryID: 2 primaryIdFromEndointStr: 2
2021-02-04T23:54:57.105Z info: 	userId: 4 secondaryId: 3 secondaryIdFromEndpointStr: 3
2021-02-04T23:54:57.105Z info: 	userId: 4 secondaryId: 1 secondaryIdFromEndpointStr: 1
2021-02-04T23:54:57.154Z info: 	userId: 1: promoteSecondary1ToPrimary
2021-02-04T23:54:57.154Z info: 	userId: 1 | P: 1->2
2021-02-04T23:54:57.155Z info: 	userId: 1 | S1: 2->1
2021-02-04T23:54:57.155Z info: 	userId: 1 | S2: 3->3
2021-02-04T23:54:57.241Z info: 	userId: 2: promoteSecondary1ToPrimary
2021-02-04T23:54:57.241Z info: 	userId: 2 | P: 3->1
2021-02-04T23:54:57.241Z info: 	userId: 2 | S1: 1->3
2021-02-04T23:54:57.241Z info: 	userId: 2 | S2: 2->2
2021-02-04T23:54:57.322Z info: 	userId: 3: promoteSecondary1ToPrimary
2021-02-04T23:54:57.322Z info: 	userId: 3 | P: 2->3
2021-02-04T23:54:57.323Z info: 	userId: 3 | S1: 3->2
2021-02-04T23:54:57.323Z info: 	userId: 3 | S2: 1->1
2021-02-04T23:54:57.414Z info: 	userId: 4: promoteSecondary1ToPrimary
2021-02-04T23:54:57.414Z info: 	userId: 4 | P: 2->3
2021-02-04T23:54:57.414Z info: 	userId: 4 | S1: 3->2
2021-02-04T23:54:57.414Z info: 	userId: 4 | S2: 1->1
2021-02-04T23:54:58.689Z info: 	Discovery Node has indexed block #81!
2021-02-04T23:54:58.884Z info: 	Discovery Node has indexed block #82!
2021-02-04T23:54:58.946Z info: 	Discovery Node has indexed block #83!
2021-02-04T23:54:59.036Z info: 	Discovery Node has indexed block #84!
2021-02-04T23:54:59.159Z info: 	userId: 1 Replica Set Info: 2, 1,3
2021-02-04T23:54:59.169Z info: 	userId: 1 Replica Set String: http://cn2_creator-node_1:4001,http://cn1_creator-node_1:4000,http://cn3_creator-node_1:4002
2021-02-04T23:54:59.169Z info: 	userId: 1 primaryID: 2 primaryIdFromEndointStr: 2
2021-02-04T23:54:59.169Z info: 	userId: 1 secondaryId: 1 secondaryIdFromEndpointStr: 1
2021-02-04T23:54:59.169Z info: 	userId: 1 secondaryId: 3 secondaryIdFromEndpointStr: 3
2021-02-04T23:54:59.170Z info: 	userId: 2 Replica Set Info: 1, 3,2
2021-02-04T23:54:59.170Z info: 	userId: 2 Replica Set String: http://cn1_creator-node_1:4000,http://cn3_creator-node_1:4002,http://cn2_creator-node_1:4001
2021-02-04T23:54:59.170Z info: 	userId: 2 primaryID: 1 primaryIdFromEndointStr: 1
2021-02-04T23:54:59.170Z info: 	userId: 2 secondaryId: 3 secondaryIdFromEndpointStr: 3
2021-02-04T23:54:59.170Z info: 	userId: 2 secondaryId: 2 secondaryIdFromEndpointStr: 2
2021-02-04T23:54:59.215Z info: 	userId: 3 Replica Set Info: 3, 2,1
2021-02-04T23:54:59.216Z info: 	userId: 3 Replica Set String: http://cn3_creator-node_1:4002,http://cn2_creator-node_1:4001,http://cn1_creator-node_1:4000
2021-02-04T23:54:59.216Z info: 	userId: 3 primaryID: 3 primaryIdFromEndointStr: 3
2021-02-04T23:54:59.216Z info: 	userId: 3 secondaryId: 2 secondaryIdFromEndpointStr: 2
2021-02-04T23:54:59.216Z info: 	userId: 3 secondaryId: 1 secondaryIdFromEndpointStr: 1
2021-02-04T23:54:59.216Z info: 	userId: 4 Replica Set Info: 3, 2,1
2021-02-04T23:54:59.217Z info: 	userId: 4 Replica Set String: http://cn3_creator-node_1:4002,http://cn2_creator-node_1:4001,http://cn1_creator-node_1:4000
2021-02-04T23:54:59.217Z info: 	userId: 4 primaryID: 3 primaryIdFromEndointStr: 3
2021-02-04T23:54:59.217Z info: 	userId: 4 secondaryId: 2 secondaryIdFromEndpointStr: 2
2021-02-04T23:54:59.217Z info: 	userId: 4 secondaryId: 1 secondaryIdFromEndpointStr: 1
2021-02-04T23:54:59.247Z info: 	userId: 1: promoteSecondary2ToPrimary
2021-02-04T23:54:59.248Z info: 	userId: 1 | P: 2->3
2021-02-04T23:54:59.248Z info: 	userId: 1 | S1: 1->1
2021-02-04T23:54:59.248Z info: 	userId: 1 | S2: 3->2
2021-02-04T23:54:59.273Z info: 	userId: 2: promoteSecondary2ToPrimary
2021-02-04T23:54:59.273Z info: 	userId: 2 | P: 1->2
2021-02-04T23:54:59.273Z info: 	userId: 2 | S1: 3->3
2021-02-04T23:54:59.273Z info: 	userId: 2 | S2: 2->1
2021-02-04T23:54:59.291Z info: 	userId: 3: promoteSecondary2ToPrimary
2021-02-04T23:54:59.292Z info: 	userId: 3 | P: 3->1
2021-02-04T23:54:59.292Z info: 	userId: 3 | S1: 2->2
2021-02-04T23:54:59.292Z info: 	userId: 3 | S2: 1->3
2021-02-04T23:54:59.309Z info: 	userId: 4: promoteSecondary2ToPrimary
2021-02-04T23:54:59.309Z info: 	userId: 4 | P: 3->1
2021-02-04T23:54:59.309Z info: 	userId: 4 | S1: 2->2
2021-02-04T23:54:59.309Z info: 	userId: 4 | S2: 1->3
2021-02-04T23:55:03.754Z info: 	Discovery Node has indexed block #85!
2021-02-04T23:55:03.870Z info: 	Discovery Node has indexed block #86!
2021-02-04T23:55:03.924Z info: 	Discovery Node has indexed block #87!
2021-02-04T23:55:04.013Z info: 	Discovery Node has indexed block #88!
2021-02-04T23:55:04.102Z info: 	userId: 1 Replica Set Info: 3, 1,2
2021-02-04T23:55:04.103Z info: 	userId: 1 Replica Set String: http://cn3_creator-node_1:4002,http://cn1_creator-node_1:4000,http://cn2_creator-node_1:4001
2021-02-04T23:55:04.103Z info: 	userId: 1 primaryID: 3 primaryIdFromEndointStr: 3
2021-02-04T23:55:04.103Z info: 	userId: 1 secondaryId: 1 secondaryIdFromEndpointStr: 1
2021-02-04T23:55:04.103Z info: 	userId: 1 secondaryId: 2 secondaryIdFromEndpointStr: 2
2021-02-04T23:55:04.226Z info: 	userId: 2 Replica Set Info: 2, 3,1
2021-02-04T23:55:04.226Z info: 	userId: 2 Replica Set String: http://cn2_creator-node_1:4001,http://cn3_creator-node_1:4002,http://cn1_creator-node_1:4000
2021-02-04T23:55:04.226Z info: 	userId: 2 primaryID: 2 primaryIdFromEndointStr: 2
2021-02-04T23:55:04.226Z info: 	userId: 2 secondaryId: 3 secondaryIdFromEndpointStr: 3
2021-02-04T23:55:04.226Z info: 	userId: 2 secondaryId: 1 secondaryIdFromEndpointStr: 1
2021-02-04T23:55:04.333Z info: 	userId: 3 Replica Set Info: 1, 2,3
2021-02-04T23:55:04.333Z info: 	userId: 3 Replica Set String: http://cn1_creator-node_1:4000,http://cn2_creator-node_1:4001,http://cn3_creator-node_1:4002
2021-02-04T23:55:04.333Z info: 	userId: 3 primaryID: 1 primaryIdFromEndointStr: 1
2021-02-04T23:55:04.333Z info: 	userId: 3 secondaryId: 2 secondaryIdFromEndpointStr: 2
2021-02-04T23:55:04.333Z info: 	userId: 3 secondaryId: 3 secondaryIdFromEndpointStr: 3
2021-02-04T23:55:04.475Z info: 	userId: 4 Replica Set Info: 1, 2,3
2021-02-04T23:55:04.475Z info: 	userId: 4 Replica Set String: http://cn1_creator-node_1:4000,http://cn2_creator-node_1:4001,http://cn3_creator-node_1:4002
2021-02-04T23:55:04.475Z info: 	userId: 4 primaryID: 1 primaryIdFromEndointStr: 1
2021-02-04T23:55:04.475Z info: 	userId: 4 secondaryId: 2 secondaryIdFromEndpointStr: 2
2021-02-04T23:55:04.475Z info: 	userId: 4 secondaryId: 3 secondaryIdFromEndpointStr: 3
2021-02-04T23:55:04.509Z info: 	userId: 1: swapSecondaries
2021-02-04T23:55:04.509Z info: 	userId: 1 | P: 3->3
2021-02-04T23:55:04.509Z info: 	userId: 1 | S1: 1->2
2021-02-04T23:55:04.509Z info: 	userId: 1 | S2: 2->1
2021-02-04T23:55:04.535Z info: 	userId: 2: swapSecondaries
2021-02-04T23:55:04.535Z info: 	userId: 2 | P: 2->2
2021-02-04T23:55:04.535Z info: 	userId: 2 | S1: 3->1
2021-02-04T23:55:04.535Z info: 	userId: 2 | S2: 1->3
2021-02-04T23:55:04.561Z info: 	userId: 3: swapSecondaries
2021-02-04T23:55:04.561Z info: 	userId: 3 | P: 1->1
2021-02-04T23:55:04.561Z info: 	userId: 3 | S1: 2->3
2021-02-04T23:55:04.561Z info: 	userId: 3 | S2: 3->2
2021-02-04T23:55:04.586Z info: 	userId: 4: swapSecondaries
2021-02-04T23:55:04.588Z info: 	userId: 4 | P: 1->1
2021-02-04T23:55:04.589Z info: 	userId: 4 | S1: 2->3
2021-02-04T23:55:04.589Z info: 	userId: 4 | S2: 3->2
2021-02-04T23:55:08.687Z info: 	Discovery Node has indexed block #89!
2021-02-04T23:55:08.779Z info: 	Discovery Node has indexed block #90!
2021-02-04T23:55:08.838Z info: 	Discovery Node has indexed block #91!
2021-02-04T23:55:08.950Z info: 	Discovery Node has indexed block #92!
2021-02-04T23:55:09.107Z info: 	userId: 1 Replica Set Info: 3, 2,1
2021-02-04T23:55:09.108Z info: 	userId: 1 Replica Set String: http://cn3_creator-node_1:4002,http://cn2_creator-node_1:4001,http://cn1_creator-node_1:4000
2021-02-04T23:55:09.116Z info: 	userId: 1 primaryID: 3 primaryIdFromEndointStr: 3
2021-02-04T23:55:09.117Z info: 	userId: 1 secondaryId: 2 secondaryIdFromEndpointStr: 2
2021-02-04T23:55:09.117Z info: 	userId: 1 secondaryId: 1 secondaryIdFromEndpointStr: 1
2021-02-04T23:55:09.120Z info: 	userId: 2 Replica Set Info: 2, 1,3
2021-02-04T23:55:09.120Z info: 	userId: 2 Replica Set String: http://cn2_creator-node_1:4001,http://cn1_creator-node_1:4000,http://cn3_creator-node_1:4002
2021-02-04T23:55:09.120Z info: 	userId: 2 primaryID: 2 primaryIdFromEndointStr: 2
2021-02-04T23:55:09.121Z info: 	userId: 2 secondaryId: 1 secondaryIdFromEndpointStr: 1
2021-02-04T23:55:09.121Z info: 	userId: 2 secondaryId: 3 secondaryIdFromEndpointStr: 3
2021-02-04T23:55:09.161Z info: 	userId: 4 Replica Set Info: 1, 3,2
2021-02-04T23:55:09.166Z info: 	userId: 4 Replica Set String: http://cn1_creator-node_1:4000,http://cn3_creator-node_1:4002,http://cn2_creator-node_1:4001
2021-02-04T23:55:09.166Z info: 	userId: 4 primaryID: 1 primaryIdFromEndointStr: 1
2021-02-04T23:55:09.166Z info: 	userId: 4 secondaryId: 3 secondaryIdFromEndpointStr: 3
2021-02-04T23:55:09.167Z info: 	userId: 4 secondaryId: 2 secondaryIdFromEndpointStr: 2
2021-02-04T23:55:09.175Z info: 	userId: 3 Replica Set Info: 1, 3,2
2021-02-04T23:55:09.175Z info: 	userId: 3 Replica Set String: http://cn1_creator-node_1:4000,http://cn3_creator-node_1:4002,http://cn2_creator-node_1:4001
2021-02-04T23:55:09.175Z info: 	userId: 3 primaryID: 1 primaryIdFromEndointStr: 1
2021-02-04T23:55:09.175Z info: 	userId: 3 secondaryId: 3 secondaryIdFromEndpointStr: 3
2021-02-04T23:55:09.175Z info: 	userId: 3 secondaryId: 2 secondaryIdFromEndpointStr: 2
```

hareeshnagaraj commented on discovery-provider/alembic/versions/5dd6a55bb738_usrm_content_nodes.py:
```diff
@@ -0,0 +1,42 @@
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('usrm_content_nodes',
```
usrm_content_nodes -> ursm_content_nodes

hareeshnagaraj commented on libs/src/services/web3Manager/index.js:
```diff
@@ -23,6 +24,34 @@ class Web3Manager {
   }
 
+  /** Browser and testing-compatible signTypedData */
+  static ethSignTypedData (web3, wallet, signatureData) {
```
this change can now be reverted to the prior state

hareeshnagaraj commented on libs/src/services/web3Manager/index.js:
```diff
@@ -74,7 +109,7 @@ class Web3Manager {
       // Lowercase the owner wallet. Consider using the checksum address.
       // See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-55.md.
-      return this.ownerWallet.toLowerCase()
+      return this.getWeb3().utils.toChecksumAddress(this.ownerWallet.toLowerCase())
```
remove this change

soulfulAdam commented:
Mergered

hareeshnagaraj commented on discovery-provider/alembic/versions/5aaa6198def8_user_primary_secondary_columns.py:
```diff
@@ -0,0 +1,30 @@
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.add_column('users', sa.Column('primary_id', sa.Integer(), nullable=True))
```
per discussion with @SidSethi adding another field replica_set_update_signer to also index the new _signer event field. this will be useful for us in the case of diagnosing any potential incorrect updates

hareeshnagaraj replied:
added

dmanjunath commented on libs/src/index.js:
```diff
@@ -107,7 +107,7 @@ class AudiusLibs {
    * @param {string | Web3 | Array<string>} providers web3 provider endpoint(s)
    */
-  static configInternalWeb3 (registryAddress, providers) {
+  static configInternalWeb3 (registryAddress, providers, privateKey) {
```
@hareeshnagaraj not sure if this solves the same problem, but @cheran-senthil passed in a wallet that could sign eth transactions like this https://github.com/AudiusProject/audius-k8s/pull/177/files#diff-6d777c865f2f6389dcaaf388a6fc08eef5466dddebd4058944271783450a2663R37-R45

hareeshnagaraj replied:
@SidSethi to confirm my understanding here - i think this actually does solve the same problem, however we want to be 'logged in' to the dataContracts web3 object for the reconfiguration operations later so unless we want to do this for every single operation requiring a signature we should keep the current pattern

hareeshnagaraj replied:
doing it this way also incurs a call to the provider with `ethSignTypedData` instead of just using sig-util

dmanjunath replied:
just so i'm clear, the purpose of this pkey is for to get the correct ownerWallet here (https://github.com/AudiusProject/audius-protocol/pull/1141/files#diff-0a02a35fbb734317785e1cd8180f6042c116e0d5e13482d6b5ac4bce16d02916R53-R58) so the ethSignTypedData has the right data here(https://github.com/AudiusProject/audius-protocol/pull/1141/files#diff-0a02a35fbb734317785e1cd8180f6042c116e0d5e13482d6b5ac4bce16d02916R135-R139)?

hareeshnagaraj replied:
yes, also required for reconfiguration submission in the upcoming work - right now is only used in addOrUpdateContentNode signed data generation. actually allows a private key to be used as the wallet in dataContracts, something this solution from cheran would not allow - the benefit here is that we don't have to make a remote web3 provider call to actually use `ethSignTypedData` (useExternalWeb3 becomes false in content node), and can just use sig-util. this is already done in web3Manager

hareeshnagaraj commented on discovery-provider/src/tasks/user_replica_set.py:
```diff
@@ -0,0 +1,283 @@
+            content_node_service_type,
+            sp_id
+        ).call()
+        logger.info(cn_endpoint_info)
```
add context to this log statement

hareeshnagaraj commented on libs/src/services/creatorNode/CreatorNodeSelection.js:
```diff
@@ -2,6 +2,15 @@ const ServiceSelection = require('../../service-selection/ServiceSelection')
 const { timeRequestsAndSortByVersion } = require('../../utils/network')
 const { CREATOR_NODE_SERVICE_NAME, DECISION_TREE_STATE } = require('./constants')
 
+let contentNodeEndpointToSpID = { }
```
comments explaining these data structures

hareeshnagaraj commented on mad-dog/src/tests/test_userReplicaSetManager.js:
```diff
@@ -0,0 +1,278 @@
+    logger.info(`userId: ${userId} | S1: ${secondaries[0]}->${newSecondaries[0]}`)
+    logger.info(`userId: ${userId} | S2: ${secondaries[1]}->${newSecondaries[1]}`)
+    let tx = await libs.updateReplicaSet(userId, newPrimary, newSecondaries)
+    await waitForBlock(libs, tx.blockNumber)
```
verify the signer here

hareeshnagaraj replied:
nvm, this is already done up here https://github.com/AudiusProject/audius-protocol/pull/1141/files#diff-3d579dcaaec0479950c9d8be459360c63da1f094dd50c24aab97ef0c3a470d86R46

hareeshnagaraj commented:
Comments from here - https://github.com/AudiusProject/audius-protocol/pull/1248 have been pushed

hareeshnagaraj commented on libs/src/services/dataContracts/userReplicaSetManagerClient.js:
```diff
@@ -0,0 +1,161 @@
+   * @param {number} oldPrimary
+   * @param {Array<number>} oldSecondaries
+   */
+  async _updateReplicaSet (userId, primary, secondaries, oldPrimary, oldSecondaries) {
```
put this at the bottom

hareeshnagaraj commented on mad-dog/src/tests/test_userReplicaSetManager.js:
```diff
@@ -0,0 +1,278 @@
+      let walletsInfo = await libs.getContentNodeWallets(spID)
+      let delegateWalletFromChain = walletsInfo.delegateOwnerWallet
+      if (wallet !== delegateWalletFromChain) {
+        throw new Error(
```
verify the ownerWallet

SidSethi commented on libs/src/services/dataContracts/userReplicaSetManagerClient.js:
```diff
@@ -0,0 +1,161 @@
+  async getUserReplicaSet (userId) {
+    const method = await this.getMethod('getUserReplicaSet', userId)
+    let currentWallet = this.web3Manager.getWalletAddress()
+    return method.call({ from: currentWallet })
```
prob we should wrap this inside a `webManager.sendCall()` function or something but w/e, works for now

hareeshnagaraj replied:
eh

SidSethi commented on libs/src/services/dataContracts/userReplicaSetManagerClient.js:
```diff
@@ -0,0 +1,161 @@
+  }
+
+  /**
+   * Return the current wallet address associated with a given spID
```
update comment to reflect two wallets being returned now

SidSethi commented on libs/src/api/user.js:
```diff
@@ -683,6 +684,47 @@ class Users extends Base {
     return pick(metadata, USER_PROPS.concat('user_id'))
   }
+
+  async _updateReplicaSet (userId, metadata) {
```
can you throw a function comment + for explicitness maybe call this `_updateReplicaSetOnChain()`

SidSethi replied:
also nit - would be cleaner/more explicit to have this function just accept the endpoint field instead of entire metadata object

hareeshnagaraj replied:
totally agreed on both counts

SidSethi commented on libs/src/api/user.js:
```diff
@@ -683,6 +684,47 @@ class Users extends Base {
+    return tx
+  }
+
+  async _retrieveSpIDFromEndpoint (endpoint) {
```
a short comment would be great

SidSethi commented on libs/src/api/user.js:
```diff
@@ -683,6 +684,47 @@ class Users extends Base {
+  }
+
+  async _retrieveSpIDFromEndpoint (endpoint) {
+    let cachedSpID = await getSpIDFromEndpoint(endpoint)
```
`getSpIDFromEndpoint` is not an async function

SidSethi commented on libs/src/api/user.js:
```diff
@@ -683,6 +684,47 @@ class Users extends Base {
+      spID = spEndpointInfo.spID
+    }
+    return spID
+  }
```
as is, a single failed web3 call in this function will short circuit the full _updateReplicaSet() call - this seems correct,just confirming also we should explicitly comment. could also think about retries but would be down to just throw

also if `spEndpointInfo` on line 721 is null/empty then spID on line 724 will also be empty, at that point i thnk we should error instead of writing this to chain

hareeshnagaraj replied:
added comment explaining error

SidSethi commented on libs/src/api/user.js:
```diff
@@ -683,6 +684,47 @@ class Users extends Base {
+    let secondaries = CreatorNode.getSecondaries(metadata.creator_node_endpoint)
+    let primarySpID = await this._retrieveSpIDFromEndpoint(primaryEndpoint)
+    let secondary1SpID = await this._retrieveSpIDFromEndpoint(secondaries[0])
+    let secondary2SpID = await this._retrieveSpIDFromEndpoint(secondaries[1])
```
nit can Promise.all these for #perf

SidSethi commented on mad-dog/src/tests/test_userReplicaSetManager.js:
```diff
@@ -0,0 +1,278 @@
+  await promoteSecondary2ToPrimary(executeAll)
+  await verifyUserReplicaSets(executeAll)
+  await swapSecondaries(executeAll)
+  await verifyUserReplicaSets(executeAll)
```
this is so nice, thanks for splitting out this way

SidSethi reviewed (commented):
mostly looks great, a bunch of commenits (lol) and a couple questions around the `updateReplicaSet` logic in `libs/src/api/user`

would be good to have a sentence in PR description that this is a back-compat release of libs for both pre- and post-contract deploy support

hareeshnagaraj commented on libs/src/services/web3Manager/index.js:
```diff
@@ -49,6 +50,12 @@ class Web3Manager {
+        this.ownerWallet = EthereumWallet.fromPrivateKey(pkeyBuffer)
+        return
+      }
+
```
how it the hedgehog wallet created if the below instructions on line 60 aren't hit?

hareeshnagaraj replied:
per offline discussion it won't be hit - there is a return statement

raymondjacobson replied:
yeah this seems ok, per our discussion that we need to rework this a bit soon

hareeshnagaraj commented on libs/examples/initAudiusLibs.js:
```diff
@@ -11,11 +11,17 @@ const ethWeb3ProviderEndpoint = 'http://localhost:8546'
   if (useExternalWeb3) {
-    const dataWeb3 = new Web3(new Web3.providers.HttpProvider(dataWeb3ProviderEndpoints[0]))
+    // const dataWeb3 = new Web3(new Web3.providers.HttpProvider(dataWeb3ProviderEndpoints[0]))
+    const dataWeb3 = new Web3(dataWeb3ProviderEndpoints[0])
```
experimentation, will revert

dmanjunath reviewed (commented):
just finished the first pass at just looking through the code, no substantive things, mostly just nits you can ignore if you don't think they're worth fixing

SidSethi commented on libs/src/services/creatorNode/CreatorNodeSelection.js:
```diff
@@ -2,6 +2,15 @@ const ServiceSelection = require('../../service-selection/ServiceSelection')
+  return contentNodeEndpointToSpID[endpoint]
+}
+
+function setEndpointToSpID (spID, endpoint) {
```
also reminder to call `setEndpoit()` in libs api user `retrieveSpIDFromEndopint()`

hareeshnagaraj commented on libs/src/services/creatorNode/CreatorNodeSelection.js:
```diff
@@ -33,16 +49,6 @@ class CreatorNodeSelection extends ServiceSelection {
     this.maxStorageUsedPercent = maxStorageUsedPercent
   }
 
-  /**
```
good point @dmanjunath pointed out the same thing

hareeshnagaraj commented on libs/src/api/user.js:
```diff
@@ -683,6 +684,58 @@ class Users extends Base {
+      )
+      // Throw if this spID is 0, indicating invalid
+      spID = spEndpointInfo.spID
+      if (spID === 0) {
```
@SidSethi tested with an invalid endpoint and the value is zero so this should be covered

SidSethi replied:
perfect

dmanjunath commented:
also can we change this default to 3. this is why i was running into that issue yesterday https://github.com/AudiusProject/audius-protocol/blob/master/service-commands/scripts/setup.js#L5

SidSethi commented on libs/src/services/dataContracts/userReplicaSetManagerClient.js:
```diff
@@ -100,11 +100,15 @@ class UserReplicaSetManagerClient extends ContractClient {
+    return {
+      primaryId: parseInt(resp.primaryId),
+      secondaryIds: resp.secondaryIds.map(x => parseInt(x))
+    }
```
this is fine but why do this? i think this breaks contract client pattern. is this just bc the og response has the numbers as strings?

hareeshnagaraj replied:
seems to follow the other datacontract clients - https://github.com/AudiusProject/audius-protocol/blob/master/libs/src/services/dataContracts/trackFactoryClient.js#L45. 

@dmanjunath found this discrepancy and is only an issue in dapp environment not mad dog - not sure why the two are different though

SidSethi reviewed (approved):
looks great, thanks for addressing all changes

dmanjunath reviewed (approved):
I think this is good to go. Tested extensively both with and without the contract and new accounts, updating accounts, modifying replica set around all seems to work
